src/undo.h

```diff
@@ -188,39 +123,165 @@ class CTxUndo
+        }
+        if (contentBitMask & AvailableSections::CEASED_CERTIFICATE_DATA)
+        {
+            ::Serialize(s, ceasedBwts, nType, nVersion);
+        }
+        return;
+    }
+
+    template<typename Stream>
+    void Unserialize(Stream& s, int nType, int nVersion)
+    {
+        ::Unserialize(s, sidechainUndoDataVersion, nType, nVersion);
+        uint8_t tmp;
+        ::Unserialize(s, tmp, nType, nVersion);
+        contentBitMask = static_cast<AvailableSections>(tmp);
+        if (contentBitMask & AvailableSections::SIDECHAIN_STATE)
+        {
+            ::Unserialize(s, prevTopCommittedCertReferencedEpoch, nType, nVersion);
+            ::Unserialize(s, prevTopCommittedCertHash,            nType, nVersion);
+            ::Unserialize(s, prevTopCommittedCertQuality,         nType, nVersion);
+            ::Unserialize(s, prevTopCommittedCertBwtAmount,       nType, nVersion);
+        }
+        if (contentBitMask & AvailableSections::MATURED_AMOUNTS)
+        {
+            ::Unserialize(s, appliedMaturedAmount, nType, nVersion);
+        }
+        if (contentBitMask & AvailableSections::LOW_QUALITY_CERT_DATA)
+        {
+            ::Unserialize(s, lowQualityBwts, nType, nVersion);
+        }
+        if (contentBitMask & AvailableSections::CEASED_CERTIFICATE_DATA)
+        {
+            ::Unserialize(s, ceasedBwts, nType, nVersion);
+        }
+        return;
     }
 
     std::string ToString() const
     {
-        std::string str;
-        str += strprintf("  immAmount %d.%08d\n", appliedMaturedAmount / COIN, appliedMaturedAmount % COIN);
-        return str;
+        std::string res;
+        res += strprintf("contentBitMask=%u\n", contentBitMask);
+        if (contentBitMask && AvailableSections::SIDECHAIN_STATE)
+        {
+            res += strprintf("prevTopCommittedCertReferencedEpoch=%d\n", prevTopCommittedCertReferencedEpoch);
+            res += strprintf("prevTopCommittedCertHash=%s\n", prevTopCommittedCertHash.ToString());
+            res += strprintf("prevTopCommittedCertQuality=%d\n", prevTopCommittedCertQuality);
+            res += strprintf("prevTopCommittedCertBwtAmount=%d.%08d\n", prevTopCommittedCertBwtAmount / COIN, prevTopCommittedCertBwtAmount % COIN);
+        }
+
+        if (contentBitMask && AvailableSections::MATURED_AMOUNTS)
+            res += strprintf("appliedMaturedAmount=%d.%08d\n", appliedMaturedAmount / COIN, appliedMaturedAmount % COIN);
+
+        res += strprintf("ceasedBwts.size()=%u\n", ceasedBwts.size());
+        for(const auto& voidCertOutput: ceasedBwts)
+           res += voidCertOutput.ToString() + "\n";
+
+        res += strprintf("lowQualityBwts.size %u\n", lowQualityBwts.size());
+        for(const auto& voidCertUndo: lowQualityBwts)
+            res += voidCertUndo.ToString() + "\n";
+
+        return res;
     }
 };
 
+inline CSidechainUndoData::AvailableSections operator | (CSidechainUndoData::AvailableSections lhs, CSidechainUndoData::AvailableSections rhs)
+{
```

Comment:
If the bitmask is typed as an uint8_t (as per comment above) we can remove these operators

Reply:
Done